docs/src/proposals/big-transactions.md

```diff
@@ -0,0 +1,186 @@
+as well as runtime handling for looking up and loading accounts from the
+on-chain indexes.
+
+#### State
+
+Once created, index accounts may not be deleted. Stored addresses should be
+append only so that once an address is stored in an index account, it may not be
+removed later.
+
+Since transactions use a u16 offset to look up addresses, index accounts can
+store up to 2^16 addresses each. Anyone may create an index account of any size
+as long as its big enough to store the necessary metadata. In addition to
+stored addresses, index accounts must also track the latest count of stored
+addresses and an authority which must be a present signer for all index
+modifications.
+
+#### Program controlled indexes
+
+If the authority of an index account is controlled by a program, more
+sophisticated indexes could be built with governance features or price curves
+for new index addresses.
+
+### Versioned Transactions
+
+In order to allow accounts to be referenced more succinctly, the structure of
+serialized transactions must be modified. This means that the new transaction
+format must be distinguished from the current transaction format.
+
+Current transactions can fit at most 19 signatures (64-bytes each) but the
+message format encodes the number of required signers as a `u8`.  Since the
+upper bit of the `u8` will never be set for a valid transaction, we can enable
+it to denote whether a transaction should be decoded with the versioned format
+or not.
+
+#### New Transaction Format
+
+```rust
+pub struct VersionedMessage {
+    /// Version of encoded message.
+    /// The max encoded version is 2^7 - 1 due to the ignored upper disambiguation bit
+    pub version: u8,
+    pub header: MessageHeader,
+    /// Number of read-only account inputs specified thru indexes
+    pub num_readonly_indexed_accounts: u8,
+    #[serde(with = "short_vec")]
+    pub account_keys: Vec<Pubkey>,
+    /// All the account indexes used by this transaction
+    #[serde(with = "short_vec")]
+    pub account_indexes: Vec<AccountIndex>,
+    pub recent_blockhash: Hash,
+    /// Compiled instructions stay the same, account indexes continue to be stored
+    /// as a u8 which means the max number of account_indexes + account_keys is 256.
+    #[serde(with = "short_vec")]
+    pub instructions: Vec<CompiledInstruction>,
+}
+
+pub struct AccountIndex {
+    pub account_key_offset: u8,
+    // 1-3 bytes used to lookup address in index account
+    pub index_account_offset: CompactU16,
+}
+```
+
+#### Size changes
+
+- Extra byte for version field
+- Extra byte for number of total account index inputs
+- Extra byte for number of readonly account index inputs
+- Most indexes will be compact and use 2 bytes + index address
+- Cost of each additional index account is ~2 bytes
+
+### Limitations
+
+- Max of 256 accounts may be specified in a transaction because u8 is used by compiled
+instructions to index into transaction message account keys.
```

Comment:
We'll have to change the interpretation of the `CompiledInstruction::accounts` to determine whether they're referencing an index into `account_keys` or `account_indexes`, right? Cleanest would be to steal the top bit again, more effective to do something like `i < 37: account_keys[i]` and `i >= 37: account_indexes[i-37]`

Reply:
I was imagining that account keys and indexes would be concatenated and indexed as if they were a single array.

Reply:
then, I think some careful reconciliation is needed as for read-only/read-write handling if the same account address is referenced from directly (old way) and indirectly (new way) from tx.

Reply:
> then, I think some careful reconciliation is needed as for read-only/read-write handling if the same account address is referenced from directly (old way) and indirectly (new way) from tx.

Good edge case. I think this case should be handled with the existing `AccountLoadedTwice` error